feat: Add CDI device support

jdon opened the pull request:
This PR allows users to use the CDI interface to pass through GPUs to their containers instead of using the existing `gpus  = "all"`.

This is needed as operating systems like [nixos](https://discourse.nixos.org/t/virtualisation-docker-enablenvidia-vs-hardware-nvidia-container-toolkit-enable/51370) are removing the support for `gpus all`.

### Changes:
  - Adds CDI (Container Device Interface) device support via a new device_requests field
  - Added ConflictsWith validation between gpus and device_requests
  - Updated documentation
  - Added tests for:
    - CDI device
    - GPU
    - Conflict validation
  - Added check in tests to skip them if a NVIDIA GPU is not found.

Junkern commented:
Thanks for your PR! I spent some minutes looking into the codebase to understand the context and your request.
I might be wrong, but what you are trying to achieve is already possible/implemented?

I think the documentation around the `gpu` parameter is wrong: 
* `Currently, only the value `all` is supported.`  => this is not correct - we pass the value of the `gpus` atttribute to the docker API exactly as the Docker CLI does.

But I suppose you tried access an nVidia GPU (https://docs.docker.com/reference/cli/docker/container/run/#gpus) and it didn't work?

Then we might be missing a small detail in the implementation

jdon commented:
Hey @Junkern, 

Thanks for the ping here, I dropped the ball on this, I've just been using a fork locally for my needs :joy:.

It would be great to get this merged! I've just fixed up the conflicts. Please let me know what changes you would like and I'm more than happy to make them.

> (In addition to this PR I will fix the gpus implementation)

What are your plans for this? Are you planning on removing `gpus` and pushing everyone down the device requests route?


(Also feel free to push commits to this branch if that's quicker / easier for you!)
